File: FBSimulatorControl/Commands/FBSimulatorLaunchCtlCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorLaunchCtlCommands.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimRuntime.h>

#import "FBSimulator+Private.h"
#import "FBSimulator.h"
#import "FBSimulatorError.h"

@interface FBSimulatorLaunchCtlCommands ()

@property (nonatomic, strong, readonly) FBSimulator *simulator;
@property (nonatomic, copy, readonly) NSString *launchctlLaunchPath;

- (instancetype)initWithSimulator:(FBSimulator *)simulator launchctlLaunchPath:(NSString *)launchctlLaunchPath;

@end

@implementation FBSimulatorLaunchCtlCommands

#pragma mark Initializers

+ (NSString *)launchCtlLaunchPathForSimulator:(FBSimulator *)simulator error:(NSError **)error
{
  NSString *path = [[simulator.device.runtime.root
    stringByAppendingPathComponent:@"bin"]
    stringByAppendingPathComponent:@"launchctl"];
  FBBinaryDescriptor *binary = [FBBinaryDescriptor binaryWithPath:path error:error];
  if (!binary) {
    return nil;
  }
  return binary.path;
}

+ (instancetype)commandsWithTarget:(FBSimulator *)target
{
  NSError *error = nil;
  NSString *launchctlLaunchPath = [self launchCtlLaunchPathForSimulator:target error:&error];
  NSAssert(launchctlLaunchPath, @"Could not find path for launchctl binary with error %@", error);
  return [[FBSimulatorLaunchCtlCommands alloc] initWithSimulator:target launchctlLaunchPath:launchctlLaunchPath];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator launchctlLaunchPath:(NSString *)launchctlLaunchPath
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;
  _launchctlLaunchPath = launchctlLaunchPath;

  return self;
}

#pragma mark Querying Services

- (FBFuture<NSString *> *)serviceNameForProcessIdentifier:(pid_t)pid
{
  NSError *error = nil;
  NSString *pattern = [NSString stringWithFormat:@"^%@\t", [NSRegularExpression escapedPatternForString:@(pid).stringValue]];
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:pattern options:0 error:&error];
  if (error) {
    return [[FBSimulatorError
      describeFormat:@"Couldn't build search pattern for '%@'", @(pid)]
      failFuture];
  }

  return [[self
    firstServiceNameAndProcessIdentifierMatching:regex]
    onQueue:self.simulator.asyncQueue map:^(NSArray<id> *tuple) {
      return [tuple firstObject];
    }];
}


- (FBFuture<NSString *> *)serviceNameForProcess:(FBProcessInfo *)process
{
  return [self serviceNameForProcessIdentifier:process.processIdentifier];
}

- (FBFuture<NSDictionary<NSString *, NSNumber *> *> *)serviceNamesAndProcessIdentifiersMatching:(NSRegularExpression *)regex
{
  return [[self
    runWithArguments:@[@"list"]]
    onQueue:self.simulator.asyncQueue fmap:^(NSString *text) {
      NSArray<NSString *> *lines = [text componentsSeparatedByCharactersInSet:NSCharacterSet.newlineCharacterSet];
      NSMutableDictionary<NSString *, NSNumber *> *mapping = [NSMutableDictionary dictionary];
      for (NSString *line in lines) {
        if (![regex firstMatchInString:line options:0 range:NSMakeRange(0, line.length)]) {
          continue;
        }
        NSError *error = nil;
        pid_t processIdentifier = 0;
        NSString *serviceName = [FBSimulatorLaunchCtlCommands extractServiceNameFromListLine:line processIdentifierOut:&processIdentifier error:&error];
        if (!serviceName) {
          return [FBControlCoreError failFutureWithError:error];
        }
        mapping[serviceName] = @(processIdentifier);
      }
      return [FBFuture futureWithResult:mapping];
    }];
}

- (FBFuture<NSArray<id> *> *)firstServiceNameAndProcessIdentifierMatching:(NSRegularExpression *)regex
{
  return [[self
    serviceNamesAndProcessIdentifiersMatching:regex]
    onQueue:self.simulator.asyncQueue fmap:^(NSDictionary<NSString *, NSNumber *> *serviceNameToProcessIdentifier) {
      if (serviceNameToProcessIdentifier.count == 0) {
        return [[FBSimulatorError
          describeFormat:@"No Matching processes for '%@'", regex.pattern ]
          failFuture];
      }
      if (serviceNameToProcessIdentifier.count > 1) {
        return [[FBSimulatorError
          describeFormat:@"Multiple Matching processes for '%@' %@", regex.pattern, [FBCollectionInformation oneLineDescriptionFromDictionary:serviceNameToProcessIdentifier]]
          failFuture];
      }
      NSString *serviceName = serviceNameToProcessIdentifier.allKeys.firstObject;
      NSNumber *processIdentifier = serviceNameToProcessIdentifier.allValues.firstObject;
      return [FBFuture futureWithResult:@[serviceName, processIdentifier]];
    }];
}

- (FBFuture<NSNumber *> *)processIsRunningOnSimulator:(FBProcessInfo *)process
{
  return [[self
    serviceNameForProcess:process]
    onQueue:self.simulator.workQueue map:^NSNumber *(NSString *result) {
      return @YES;
    }];
}

- (FBFuture<NSDictionary<NSString *, id> *> *)listServices
{
  return [[self
    runWithArguments:@[@"list"]]
    onQueue:self.simulator.asyncQueue fmap:^(NSString *text) {
      NSArray<NSString *> *lines = [text componentsSeparatedByCharactersInSet:NSCharacterSet.newlineCharacterSet];
      if (lines.count < 2) {
        return [[FBSimulatorError
          describeFormat:@"Insufficient number of lines from output '%@'", text]
          failFuture];
      }
      lines = [lines subarrayWithRange:NSMakeRange(1, lines.count -1)];

      NSMutableDictionary<NSString *, id> *services = [NSMutableDictionary dictionary];
      for (NSString *line in lines) {
        if (line.length == 0) {
          continue;
        }
        pid_t processIdentifier = -1;
        NSError *error = nil;
        NSString *serviceName = [FBSimulatorLaunchCtlCommands extractServiceNameFromListLine:line processIdentifierOut:&processIdentifier error:&error];
        if (!serviceName) {
          return [FBSimulatorError failFutureWithError:error];
        }
        services[serviceName] = processIdentifier > 0 ? @(processIdentifier) : NSNull.null;
      }
      return [FBFuture futureWithResult:[services copy]];
    }];
}

#pragma mark Manipulating Services

- (FBFuture<NSString *> *)stopServiceWithName:(NSString *)serviceName
{
  return [[self
    runWithArguments:@[@"stop", serviceName]]
    rephraseFailure:@"Failed to stop service '%@'", serviceName];
}

- (FBFuture<NSString *> *)startServiceWithName:(NSString *)serviceName
{
  return [[self
    runWithArguments:@[@"start", serviceName]]
    rephraseFailure:@"Failed to start service '%@'", serviceName];
}

#pragma mark Helpers

+ (nullable NSString *)extractApplicationBundleIdentifierFromServiceName:(NSString *)serviceName
{
  NSRegularExpression *regex = self.regularExpressionForServiceNameToBundleID;
  NSTextCheckingResult *result = [regex firstMatchInString:serviceName options:0 range:NSMakeRange(0, serviceName.length)];
  if (!result) {
    return nil;
  }
  NSRange range = [result rangeAtIndex:1];
  return [serviceName substringWithRange:range];
}

#pragma mark Private

+ (NSRegularExpression *)regularExpressionForServiceNameToBundleID
{
  static dispatch_once_t onceToken;
  static NSRegularExpression *regex;
  dispatch_once(&onceToken, ^{
    NSError *error = nil;
    regex = [NSRegularExpression regularExpressionWithPattern:@"UIKitApplication:([^\\[]*).*" options:NSRegularExpressionDotMatchesLineSeparators error:&error];
    NSCAssert(regex, @"Should be able to compile regex %@", error);
  });
  return regex;
}

+ (NSString *)extractServiceNameFromListLine:(NSString *)line processIdentifierOut:(pid_t *)processIdentifierOut error:(NSError **)error
{
  NSArray<NSString *> *words = [line componentsSeparatedByCharactersInSet:NSCharacterSet.whitespaceCharacterSet];
  if (words.count != 3) {
    return [[FBSimulatorError
      describeFormat:@"Output does not have exactly three words: %@", [FBCollectionInformation oneLineDescriptionFromArray:words]]
      fail:error];
  }
  NSString *serviceName = [words lastObject];
  NSString *processIdentifierString = [words firstObject];
  if ([processIdentifierString isEqualToString:@"-"]) {
    if (processIdentifierOut) {
      *processIdentifierOut = -1;
    }
    return serviceName;
  }

  NSInteger processIdentifierInteger = [processIdentifierString integerValue];
  if (processIdentifierInteger < 1) {
    return [[FBSimulatorError
      describeFormat:@"Expected a process identifier as first word, but got %@ from %@", processIdentifierString, [FBCollectionInformation oneLineDescriptionFromArray:words]]
      fail:error];
  }
  if (processIdentifierOut) {
    *processIdentifierOut = (pid_t) processIdentifierInteger;
  }

  return serviceName;
}

- (FBFuture<NSString *> *)runWithArguments:(NSArray<NSString *> *)arguments
{
  // Construct a Launch Configuration for launchctl we'll use the 'list' command.
  FBProcessSpawnConfiguration *launchConfiguration = [[FBProcessSpawnConfiguration alloc]
    initWithLaunchPath:self.launchctlLaunchPath
    arguments:arguments
    environment:@{}
    io:FBProcessIO.outputToDevNull
    mode:FBProcessSpawnModeDefault];

  // Spawn and get the output
  return [FBProcessSpawnCommandHelpers launchConsumingStdout:launchConfiguration withCommands:self.simulator];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorLaunchCtlCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
@preconcurrency import Foundation

// swiftlint:disable force_try

public final class FBSimulatorLaunchCtlCommands: NSObject {

  // MARK: - Properties

  private let simulator: FBSimulator
  private let launchctlLaunchPath: String

  // MARK: - Initializers

  private class func launchCtlLaunchPath(for simulator: FBSimulator) throws -> String {
    let path = (simulator.device.runtime.root as NSString)
      .appendingPathComponent("bin")
      .appending("/launchctl")
    let binary = try FBBinaryDescriptor.binary(withPath: path)
    return binary.path
  }

  public class func commands(with simulator: FBSimulator) -> FBSimulatorLaunchCtlCommands {
    let launchctlLaunchPath = try! launchCtlLaunchPath(for: simulator)
    return FBSimulatorLaunchCtlCommands(simulator: simulator, launchctlLaunchPath: launchctlLaunchPath)
  }

  private init(simulator: FBSimulator, launchctlLaunchPath: String) {
    self.simulator = simulator
    self.launchctlLaunchPath = launchctlLaunchPath
    super.init()
  }

  // MARK: - Services

  fileprivate func serviceName(forProcessIdentifier pid: pid_t) async throws -> String {
    let pattern = "^\(NSRegularExpression.escapedPattern(for: "\(pid)"))\t"
    guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
      throw FBSimulatorError.describe("Couldn't build search pattern for '\(pid)'").build()
    }
    let (serviceName, _) = try await firstServiceNameAndProcessIdentifier(matching: regex)
    return serviceName
  }

  fileprivate func serviceNamesAndProcessIdentifiers(matching regex: NSRegularExpression) async throws -> [String: NSNumber] {
    let text = try await run(.list)
    let lines = text.components(separatedBy: .newlines)
    var mapping: [String: NSNumber] = [:]
    for line in lines {
      if regex.firstMatch(in: line, options: [], range: NSRange(location: 0, length: line.count)) == nil {
        continue
      }
      var processIdentifier: pid_t = 0
      guard let serviceName = try? FBSimulatorLaunchCtlCommands.extractServiceName(fromListLine: line, processIdentifierOut: &processIdentifier) else {
        // If extraction fails, skip the line
        continue
      }
      mapping[serviceName] = NSNumber(value: processIdentifier)
    }
    return mapping
  }

  fileprivate func firstServiceNameAndProcessIdentifier(matching regex: NSRegularExpression) async throws -> (String, pid_t) {
    let serviceNameToProcessIdentifier = try await serviceNamesAndProcessIdentifiers(matching: regex)
    guard let (serviceName, processIdentifier) = serviceNameToProcessIdentifier.first else {
      throw FBSimulatorError.describe("No Matching processes for '\(regex.pattern)'").build()
    }
    if serviceNameToProcessIdentifier.count > 1 {
      throw FBSimulatorError.describe("Multiple Matching processes for '\(regex.pattern)' \(FBCollectionInformation.oneLineDescription(from: serviceNameToProcessIdentifier))").build()
    }
    return (serviceName, processIdentifier.int32Value)
  }

  fileprivate func listServices() async throws -> [String: Any] {
    let text = try await run(.list)
    let lines = text.components(separatedBy: .newlines)
    if lines.count < 2 {
      throw FBSimulatorError.describe("Insufficient number of lines from output '\(text)'").build()
    }
    var services: [String: Any] = [:]
    for (serviceName, processIdentifier) in Self.serviceMap(fromListOutput: text) {
      services[serviceName] = processIdentifier > 0 ? NSNumber(value: processIdentifier) : NSNull()
    }
    return services
  }

  fileprivate func stopService(withName serviceName: String) async throws -> String {
    do {
      return try await run(.stop(serviceName: serviceName))
    } catch {
      throw FBSimulatorError.describe("Failed to stop service '\(serviceName)'")
        .caused(by: error as NSError)
        .build()
    }
  }

  fileprivate func startService(withName serviceName: String) async throws -> String {
    do {
      return try await run(.start(serviceName: serviceName))
    } catch {
      throw FBSimulatorError.describe("Failed to start service '\(serviceName)'")
        .caused(by: error as NSError)
        .build()
    }
  }

  // MARK: - Helpers

  @objc
  public class func extractApplicationBundleIdentifier(fromServiceName serviceName: String) -> String? {
    guard let marker = serviceName.range(of: "UIKitApplication:") else {
      return nil
    }
    return String(serviceName[marker.upperBound...].prefix { $0 != "[" })
  }

  // MARK: - Private

  private class func extractServiceName(fromListLine line: String, processIdentifierOut: inout pid_t) throws -> String {
    let words = line.components(separatedBy: .whitespaces)
    guard words.count == 3, let processIdentifierString = words.first, let serviceName = words.last else {
      throw FBSimulatorError.describe("Output does not have exactly three words: \(FBCollectionInformation.oneLineDescription(from: words))").build()
    }
    if processIdentifierString == "-" {
      processIdentifierOut = -1
      return serviceName
    }

    let processIdentifierInteger = Int(processIdentifierString) ?? 0
    guard processIdentifierInteger >= 1 else {
      throw FBSimulatorError.describe("Expected a process identifier as first word, but got \(processIdentifierString) from \(FBCollectionInformation.oneLineDescription(from: words))").build()
    }
    processIdentifierOut = pid_t(processIdentifierInteger)
    return serviceName
  }

  // Parses `launchctl list` output into a service-name -> pid map: a pid of -1 marks a
  // loaded-but-not-running service (the "-" placeholder), >= 1 is a live process. The header
  // row and any malformed (non-three-column) lines are skipped. Static and internal so the
  // parsing can be unit-tested without spawning launchctl; see FBSimulatorLaunchCtlCommandsTests.
  static func serviceMap(fromListOutput text: String) -> [String: pid_t] {
    var services: [String: pid_t] = [:]
    for line in text.components(separatedBy: .newlines) {
      if line.isEmpty {
        continue
      }
      var processIdentifier: pid_t = -1
      guard let serviceName = try? extractServiceName(fromListLine: line, processIdentifierOut: &processIdentifier) else {
        continue
      }
      services[serviceName] = processIdentifier
    }
    return services
  }

  // The closed set of launchctl operations this command issues. Modelling them as an enum keeps argv
  // construction in one place and makes the operation set exhaustive, so a new operation cannot be
  // added without routing through `run`.
  enum Command {
    case list
    case stop(serviceName: String)
    case start(serviceName: String)

    var arguments: [String] {
      switch self {
      case .list:
        return ["list"]
      case let .stop(serviceName):
        return ["stop", serviceName]
      case let .start(serviceName):
        return ["start", serviceName]
      }
    }

    var exitCodePolicy: ExitCodePolicy {
      switch self {
      case .list:
        return .require([0])
      case .stop:
        // launchctl returns ESRCH (3) when the service is not running; for stop that is an idempotent
        // no-op. Any other non-zero is a genuine failure to stop a running service.
        return .require([0, 3])
      case .start:
        // For start, ESRCH (3) means there is no such service to start — a genuine failure, not a
        // no-op — so only a 0 exit is success.
        return .require([0])
      }
    }
  }

  private func run(_ command: Command) async throws -> String {
    let output = try await simulator.launchProcessConsumingOutput(launchPath: launchctlLaunchPath, arguments: command.arguments)
    return try FBSimulatorLaunchCtlCommands.stdout(orThrowFrom: output, command: command, logger: simulator.logger)
  }

  // Internal for unit-test coverage of the exit-code handling; see FBSimulatorLaunchCtlCommandsTests.
  static func stdout(orThrowFrom output: FBInSimulatorToolOutput, command: Command, logger: (any FBControlCoreLogger)?) throws -> String {
    if output.exitCode != 0 {
      let stderr = String(data: output.stderr, encoding: .utf8) ?? ""
      guard command.exitCodePolicy.accepts(output.exitCode) else {
        throw FBSimulatorError.describe("launchctl \(command.arguments.joined(separator: " ")) failed with exit code \(output.exitCode): \(stderr)").build()
      }
      logger?.log("launchctl \(command.arguments.joined(separator: " ")) exited with code \(output.exitCode): \(stderr)")
    }
    return String(data: output.stdout, encoding: .utf8) ?? ""
  }
}

// MARK: - FBSimulator+LaunchCtlCommands

extension FBSimulator: LaunchCtlCommands {

  public func serviceName(forProcessIdentifier pid: pid_t) async throws -> String {
    try await launchCtlCommands().serviceName(forProcessIdentifier: pid)
  }

  public func serviceName(forProcess process: FBProcessInfo) async throws -> String {
    try await launchCtlCommands().serviceName(forProcessIdentifier: process.processIdentifier)
  }

  public func serviceNamesAndProcessIdentifiers(matching regex: NSRegularExpression) async throws -> [String: NSNumber] {
    try await launchCtlCommands().serviceNamesAndProcessIdentifiers(matching: regex)
  }

  public func firstServiceNameAndProcessIdentifier(matching regex: NSRegularExpression) async throws -> (serviceName: String, processIdentifier: pid_t) {
    try await launchCtlCommands().firstServiceNameAndProcessIdentifier(matching: regex)
  }

  public func processIsRunning(onSimulator process: FBProcessInfo) async throws -> Bool {
    _ = try await launchCtlCommands().serviceName(forProcessIdentifier: process.processIdentifier)
    return true
  }

  public func listServices() async throws -> [String: Any] {
    try await launchCtlCommands().listServices()
  }

  public func stopService(withName serviceName: String) async throws -> String {
    try await launchCtlCommands().stopService(withName: serviceName)
  }

  public func startService(withName serviceName: String) async throws -> String {
    try await launchCtlCommands().startService(withName: serviceName)
  }
}
